crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,320 @@
+//! Middleware to handle `s3://` URLs to pull artifacts from an S3 bucket
+use std::path::PathBuf;
+
+use async_trait::async_trait;
+use aws_config::BehaviorVersion;
+use aws_sdk_s3::presigning::{PresignedRequest, PresigningConfig};
+use reqwest::{Request, Response};
+use reqwest_middleware::{Middleware, Next, Result as MiddlewareResult};
+use url::Url;
+
+/// S3 middleware to authenticate requests.
+pub struct S3Middleware {
+    config_file: Option<PathBuf>,
+    profile: Option<String>,
+    force_path_style: Option<bool>,
+    expiration: std::time::Duration,
+}
+
+async fn create_client(
+    config_file: Option<PathBuf>,
+    profile: Option<String>,
+    force_path_style: Option<bool>,
+) -> aws_sdk_s3::Client {
+    let mut aws_config_builder = aws_config::defaults(BehaviorVersion::latest());
+    if let Some(config_file) = config_file {
+        let mut builder = aws_runtime::env_config::file::EnvConfigFiles::builder();
+        builder = builder.with_file(
+            aws_runtime::env_config::file::EnvConfigFileKind::Config,
+            config_file,
+        );
+        let env_config_files = builder.build();
+        aws_config_builder = aws_config_builder.profile_files(env_config_files);
+    }
+
+    if let Some(profile) = profile {
+        aws_config_builder = aws_config_builder.profile_name(profile);
+    };
+    let sdk_config = aws_config_builder.load().await;
+
+    let mut builder = aws_sdk_s3::config::Builder::from(&sdk_config);
+    if let Some(force_path_style) = force_path_style {
+        builder = builder.force_path_style(force_path_style);
+    };
+    let s3_config = builder.build();
+
+    let client: aws_sdk_s3::Client = aws_sdk_s3::Client::from_conf(s3_config);
+    client
+}
+
+impl S3Middleware {
+    /// Create a new S3 middleware.
+    pub fn new(
+        config_file: Option<PathBuf>,
+        profile: Option<String>,
+        force_path_style: Option<bool>,
+    ) -> Self {
+        Self {
+            config_file,
+            profile,
+            force_path_style,
+            expiration: std::time::Duration::from_secs(300),
+        }
+    }
+
+    /// Generate a presigned S3 `GetObject` request.
+    async fn generate_presigned_s3_request(
+        &self,
+        bucket_name: &str,
+        key: &str,
+    ) -> MiddlewareResult<PresignedRequest> {
+        let client = create_client(
```

Comment:
if this wasn't async, we could also implement it in `S3iddleware::new` :/
now we need to instantiate this on every call which isn't ideal